### DISCUSSION TITLE
BEST PRACTICES for paginate using Prisma ORM + Advanced types (include, select...)

### DISCUSSION THREAD

**Discussion by felinto-dev** [prisma/prisma#7770] [Q&A]
Guys, I really love the amazing work Prisma has done and I know this work could be improved if:

1. findMany should have the option to return the count of all objects, even using the skip and take attribute. This would be useful for paginate and would avoid this:

```ts
async paginate(params: Prisma.LessonFindManyArgs): Promise<{ items: Lesson[]; total: number }> {
  const [items, total] = await Promise.all([
    this.prismaService.lesson.findMany(params),
    this.prismaService.lesson.count({ where: params.where }),
  ]);

  return { items, total };
}
```
Source: https://github.com/prisma/prisma/discussions/6501

2. Improve typing for "where" parameters
An issue has already been created about this here:
https://github.com/prisma/prisma/issues/3372

And a PR about that:
https://github.com/prisma/prisma/pull/5186/commits/796409e1f7f553c3f3760a96903a9b38c0c277f7

However, I still facing issues trying to do it:
https://github.com/prisma/prisma/discussions/7769
https://www.prisma.io/docs/concepts/components/prisma-client/advanced-type-safety/operating-against-partial-structures-of-model-types

I believe there is a huge gulf between Prisma paging documentation and real life.

Please, anyone who has a public GitHub project that has resolved these issues and created more "readable and elegant" code, please paste the link here in this thread.

This would help me to refactor my code following best practices, as well as others that might have to face the same difficulties.

**Comment by ryands17** [comment]
@felinto-dev 👋 

For point **1** we already have a feature request for the same here so you can add a 👍 and follow this for more updates. As for the workaround, this would be much better than a `Promise.all`.

```ts
async paginate(params: Prisma.LessonFindManyArgs): Promise<{ items: Lesson[]; total: number }> {
  const [items, total] = await this.prismaService.$transaction([
    this.prismaService.lesson.findMany(params),
    this.prismaService.lesson.count({ where: params.where }),
  ]);

  return { items, total };
}
```

As for point **2** I will answer it in the same discussion.

> I believe there is a huge gulf between Prisma paging documentation and real life.

I don't think that's the case.

**Reply by felinto-dev** [reply]
Hello!

Could you send me the link?

> For point 1 we already have a feature request for the same here so you can add a 👍 and follow this for more updates.

**Reply by ryands17** [reply]
https://github.com/prisma/prisma/issues/7550

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
